tools/probes/probes_main.c

```diff
@@ -182,39 +183,44 @@ void finalize_wave_files(struct wave_files *files)
 	}
 }
 
-int validate_data_packet(struct probe_data_packet *data_packet)
+int validate_data_packet(struct probe_data_packet *packet)
 {
-	uint32_t received_crc;
-	uint32_t calc_crc;
-
-	received_crc = data_packet->checksum;
-	data_packet->checksum = 0;
-	calc_crc = crc32(0, (char *)data_packet, sizeof(*data_packet));
-
-	if (received_crc == calc_crc) {
-		return 0;
-	} else {
-		fprintf(stderr, "error: data packet for buffer %d is not valid: crc32: %d/%d\n",
-			data_packet->buffer_id, calc_crc, received_crc);
+	uint64_t *checksump;
+	uint64_t sum;
+
+	sum = (uint32_t) (packet->sync_word +
+			  packet->buffer_id  +
+			  packet->format +
+			  packet->timestamp_high +
+			  packet->timestamp_low +
+			  packet->data_size_bytes);
+
+	checksump = (uint64_t *) (packet->data + packet->data_size_bytes);
+
+	if (sum != *checksump) {
+		fprintf(stderr, "Checksum error 0x%016lx != 0x%016lx\n", sum, *checksump);
 		return -EINVAL;
 	}
+
+	return 0;
 }
 
 int process_sync(struct probe_data_packet *packet, uint8_t **w_ptr, uint32_t *total_data_to_copy)
 {
 	struct probe_data_packet *temp_packet;
 
-	/* request to copy data_size from probe packet */
-	*total_data_to_copy = packet->data_size_bytes;
+	/* request to copy data_size from probe packet and 64-bit checksum */
+	*total_data_to_copy = packet->data_size_bytes + sizeof(uint64_t);
 
 	if (packet->data_size_bytes > PACKET_MAX_SIZE) {
```

Comment:
Oops... this is a bug, should be (*total_data_to_copy > PACKET_MAX_SIZE).

Reply:
Actually even the original was buggy.... the condition should be:
(sizeof(struct probe_data_packet) + *total_data_to_copy > PACKET_MAX_SIZE)

And this still not that great. Its reallocing the packet time and time again, every time packet is bigger than PACKET_MAX_SIZE. But this should be harmless waste of cycles.